analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs

```diff
@@ -128,14 +130,127 @@ public static SyntaxNode WalkUpParentheses(this SyntaxNode node)
+                .TakeWhile(x => x.IsAnyKind(
+                    SyntaxKind.Argument,
+                    SyntaxKindEx.TupleExpression,
+                    SyntaxKindEx.SingleVariableDesignation,
+                    SyntaxKindEx.ParenthesizedVariableDesignation,
+                    SyntaxKindEx.DiscardDesignation,
+                    SyntaxKindEx.DeclarationExpression))
+                .LastOrDefault();
+            if ((TupleExpressionSyntaxWrapper.IsInstance(thisSide) || DeclarationExpressionSyntaxWrapper.IsInstance(thisSide))
+                && thisSide.Parent is AssignmentExpressionSyntax assignment)
+            {
+                var otherSide = assignment switch
+                {
+                    { Left: { } left, Right: { } right } when left.Equals(thisSide) => right,
+                    { Left: { } left, Right: { } right } when right.Equals(thisSide) => left,
+                    _ => null,
+                };
+                if (TupleExpressionSyntaxWrapper.IsInstance(otherSide) || DeclarationExpressionSyntaxWrapper.IsInstance(otherSide))
+                {
+                    var indexAndCount = IndexAndCountOfNesting(node);
+                    return FindMatchingNestedNode(indexAndCount, otherSide);
+                }
+            }
+
+            return null;
+
+            static Stack<IndexCountPair> IndexAndCountOfNesting(SyntaxNode node)
+            {
+                Stack<IndexCountPair> indexAndCount = new();
+                while (TupleExpressionSyntaxWrapper.IsInstance(node?.Parent) || ParenthesizedVariableDesignationSyntaxWrapper.IsInstance(node?.Parent))
+                {
+                    if (TupleExpressionSyntaxWrapper.IsInstance(node.Parent))
+                    {
+                        var parentTuple = (TupleExpressionSyntaxWrapper)node.Parent;
+                        indexAndCount.Push(new(parentTuple.Arguments.IndexOf((ArgumentSyntax)node), parentTuple.Arguments.Count));
+                        node = parentTuple.SyntaxNode.Parent;
+                    }
+                    else if (ParenthesizedVariableDesignationSyntaxWrapper.IsInstance(node.Parent))
+                    {
+                        var parentDesignation = (ParenthesizedVariableDesignationSyntaxWrapper)node.Parent;
+                        indexAndCount.Push(new(parentDesignation.Variables.IndexOf((VariableDesignationSyntaxWrapper)node), parentDesignation.Variables.Count));
+                        node = parentDesignation.SyntaxNode;
+                    }
+
+                    if (DeclarationExpressionSyntaxWrapper.IsInstance(node.Parent)
+                        && node is { Parent.Parent: ArgumentSyntax { } argument })
+                    {
+                        node = argument;
+                    }
+                }
+                return indexAndCount;
+            }
+
+            static SyntaxNode FindMatchingNestedNode(Stack<IndexCountPair> indexAndCount, SyntaxNode node)
```

Comment:
Apologies for the delayed comment: 
It would be clearer if you'd extract `IndexCountPair` into a private record.
It would be more verbose, but IMO it will help the reader to understand what is an index and what is count, which is not easy otherwise (because you can give more descriptive names to the properties).

Reply:
+1 for records